File: 04-dmoj-ccc14s3/program.c
#include <stdio.h>
#include <stdbool.h>

#define MAX_N 100000
// cars holds the mountain top queue, cars[carsTop - 1] is always the
// next car free to move (either to the branch or straight to the lake)
int cars[MAX_N];
int carsTop;

// branch acts as a stack, cars can only enter and exit it in
// last-in-first-out order, matching the "cannot overtake" rule
int branch[MAX_N];
int branchTop;

// checks whether the cars can be sent to the lake in order 1 to n
bool possible(int n) {
    branchTop = 0;
    // lake tracks the next car number we need to send to the lake

    for (int lake = 1; lake <= n; lake++) {
        if (branchTop > 0 && branch[branchTop - 1] == lake) {
            // the car we need is already waiting at the top of the
            // branch, so just send it straight to the lake
            branchTop--;
        } else if (carsTop > 0 && cars[carsTop - 1] == lake) {
            // the car we need is right at the front of the mountain
            // top, so it can skip the branch entirely
            carsTop--;
        } else {
            // the car we need is neither at the front of the mountain
            // top nor the top of the branch, so we have to push cars
            // from the mountain top onto the branch until we either
            // find it or run out of cars to push
            while (carsTop > 0 && cars[carsTop - 1] != lake) {
                branch[branchTop] = cars[carsTop - 1];
                branchTop++;
                carsTop--;
            }
            if (carsTop == 0) {
                // ran out of cars to push and never found the one we
                // needed, so the required order is impossible
                return false;
            } else {
                // found the car we needed at the front of the mountain
                // top after pushing the others, send it to the lake
                carsTop--;
            }
        }
    }
    return true;
}

int main(void) {
    int t;
    scanf("%d", &t);

    while (t--) {
        int n;
        scanf("%d", &n);
        // read the cars top to bottom as given, since cars[carsTop - 1]
        // needs to represent the car that's free to move first, we
        // fill the array in the same order it was given and treat the
        // last element read as the top of the "stack"
        for (int i = 0; i < n; i++) {
            scanf("%d", &cars[i]);
        }
        carsTop = n;

        if (possible(n)) {
            printf("Y\n");
        } else {
            printf("N\n");
        }
    }

    return 0;
}

File: 04-dmoj-ccc14s3/program.py
# Simulates trying to drive the cars down the hill in their required order.
def possible(cars, n):
    # The branch is a last-in-first-out structure,
    # hence we use it as a stack.
    branch = []
    # We test trying to move out every car from `1` to `n` strictly in that order.
    for lake in range(1, n+1):
        # If the car is on the edge of the branch or the original mountain,
        # then it is possible to move this car out in order.
        if branch and branch[-1] == lake:
            branch.pop()
        elif cars and cars[-1] == lake:
            cars.pop()
        else:
            # Otherwise, the only way you can attempt to move out your desired car
            # is to move cars from the mountain into the branch until you find the one you want.
            while cars and cars[-1] != lake:
                branch.append(cars.pop())
            # If you cannot find the car in the process of doing this, then it is impossible.
            if len(cars) == 0:
                return False
            cars.pop()
    return True

t = int(input())
for _ in range(t):
    n = int(input())
    # The cars are inputted as they appear reading from the top down the mountain.
    # No order reversal is needed.
    cars = []
    for _ in range(n):
        cars.append(int(input()))
    print('Y' if possible(cars, n) else 'N')
